### DISCUSSION TITLE
.zsh

### DISCUSSION THREAD

**Discussion by MalikRumi** [nodejs/help#4605] [Q&A]
I was following your documentation to install nvm.  However, I changed the reference to bash at the end to zsh, since that's what I use on my mac. I was a little surprised at the tone of the response I got back. It certainly felt condescending, even disapproving. Given that so many of us Mac users are out here, and have become used to substituting zsh for instructions that reference bash, or neither, it is to be expected that a lot of us will do exactly what I did. In fact, the presence of this -- 'error message' --  strongly suggests that it happens quite frequently or you would not have put this tone deaf message into the response. That being so, you and your users would be well served by saying so **in the documentation**, right along side the sample code you give for installation. | Second, this restriction raises questions, at least for me. Are we to expect bugs and other issues with nvm - or node, npm, typescript (maintained by Microsoft... is *that* an issue?), and the rest of your ecosystem, because we have zsh? Can you list those out for us, along with workarounds?   I already know they aren't in the docs. A search for zsh only got three hits, all of which were on the same page, and two in the same paragraph. So, since you are so explicit in your messaging, you should also be explicit as to why you have this preference, and what consequences zsh users might expect from having the wrong shell. | Finally, if you are going to argue that nvm is a different app with different maintainers, that won't impress me. These are your docs I'm looking at, and since you recommend nvm you ought to know what they're doing. Besides, my request for change was about your docs, not theirs.

The fun continues…
Last login: Tue Nov 26 20:16:11 on ttys000
https://nodejs.org/en/download/package-manager/all#nvm 
_“On Unix / OS X systems Node.js built from source can be installed using [nvm](https://github.com/creationix/nvm) by installing into the location that nvm expects:”_
Apparently, the values to these constants are known to nvm, and need not be known by me. I see no other way of interpreting this command we are told to put into the command line. However:
```
malikarumi@Maliks-MacBook-Pro ~ % env VERSION=`python tools/getnodeversion.py` make install DESTDIR=`nvm_version_path v$VERSION` PREFIX=""
zsh: command not found: python
make: *** No rule to make target `install'.  Stop.
malikarumi@Maliks-MacBook-Pro ~ % which python
python not found
malikarumi@Maliks-MacBook-Pro ~ % which python3
/usr/bin/python3
malikarumi@Maliks-MacBook-Pro ~ % python3 --version
Python 3.9.6
malikarumi@Maliks-MacBook-Pro ~ % env VERSION=`python3 tools/getnodeversion.py` make install DESTDIR=`nvm_version_path v$VERSION` PREFIX=""
/Library/Developer/CommandLineTools/usr/bin/python3: can't open file '/Users/malikarumi/tools/getnodeversion.py': [Errno 2] No such file or directory
make: *** No rule to make target `install'.  Stop.
malikarumi@Maliks-MacBook-Pro ~ %
``` 
Yes, I see the error where ‘python3’ is concatenated to the ‘tools’ path. My paste of your code has the space between python3 and ‘tools’. I found this: https://stackoverflow.com/questions/45801457/node-js-python-not-found-exception-due-to-node-sass-and-node-gyp
**Surely** your install is not still dependent on Python 2.7? But the first line of the code on your repo is “from __future__ import print_function”, a sure sign of Python 2.7.
https://github.com/nodejs/node/blob/main/tools/getnodeversion.py 
tools/getnodeversion.py
You do know Apple stopped shipping Python 2 a long time ago, don’t you? A simple question: How am I supposed to get this thing working on a Mac? If your answer is homebrew, why tell mac users to do it this way in your docs? Do your docs need an upgrade?
All this, and I wasn’t interesting in node to begin with. I just need it for sass.

**Comment by hernandemonteiro** [comment]
Your frustration is understandable, especially given the challenges posed by unclear documentation and conflicting expectations for tools like `nvm` and `node`. Let’s break down the key issues and provide actionable solutions, along with suggestions for improvement that could benefit the community:

---

### 1. **NVM and Zsh Compatibility**  
   - **Issue:** Documentation seems biased toward Bash, causing confusion for Zsh users.  
   - **Actionable Fix:**  
     To ensure `nvm` works with Zsh:  
     1. Add the NVM initialization code to your `.zshrc` file:  
        ```bash
        export NVM_DIR="$HOME/.nvm"
        [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh" # This loads nvm
        [ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion" # This loads nvm bash_completion
        ```
     2. Reload your shell or source your `.zshrc` file:  
        ```bash
        source ~/.zshrc
        ```
     This step ensures Zsh properly sources `nvm` and avoids confusion in shell-dependent behavior.

   - **Suggestion for Docs Improvement:**  
     The Node.js installation docs could explicitly acknowledge Zsh and provide example setup instructions. This acknowledgment would validate users' experiences and prevent avoidable errors.

---

### 2. **Python Dependency Confusion**
   - **Issue:** The outdated reliance on Python 2.x is a pain point, especially for macOS users where Python 2.x has been deprecated.  
   - **Actionable Fix:**  
     If Python 3.x is available on your system, ensure `python` points to `python3`:  
     ```bash
     alias python=python3
     ```  
     Add this alias to your shell configuration file (e.g., `.zshrc` or `.bashrc`). Alternatively, run:  
     ```bash
     brew install pyenv
     pyenv global 3.x.x  # Replace '3.x.x' with your preferred Python 3 version.
     ```  

   - **Problem in Node.js Repo:**  
     The script in question (`tools/getnodeversion.py`) seems to assume the presence of Python 2.x. You are correct that this is outdated and should be modernized to work seamlessly with Python 3.x.

   - **Suggestion for Docs Improvement:**  
     Update the documentation to address this issue explicitly. If Python is required, the docs should:  
       - Specify the required Python version.
       - Offer clear guidance on how to install or alias the correct Python interpreter.

---

### 3. **General Installation Issues on macOS**  
   - **Problem:** The recommended installation steps don’t always account for macOS-specific challenges.  
   - **Best Practice for macOS Users:**  
     Instead of the manual `nvm` installation, consider using Homebrew for simplicity:  
     ```bash
     brew install nvm
     mkdir ~/.nvm
     export NVM_DIR="$HOME/.nvm"
     echo 'export NVM_DIR="$HOME/.nvm"' >>~/.zshrc
     echo '[ -s "$(brew --prefix nvm)/nvm.sh" ] && \. "$(brew --prefix nvm)/nvm.sh"' >>~/.zshrc
     source ~/.zshrc
     ```
     This approach resolves most pathing and dependency issues automatically.

---

### 4. **Tone and Responsiveness in Documentation**
   - **Community Feedback:**  
     The tone in technical responses matters, and phrasing perceived as dismissive or overly critical can alienate users. The team maintaining these docs should strive for a tone that is helpful and inclusive.
   - **Recommendation for Documentation:**  
     Provide disclaimers or notes that address common issues, such as:  
       - Acknowledgment of Zsh as a common shell for macOS users.  
       - Alternative instructions for Python 3.x.  
       - Clear Homebrew-based installation instructions for macOS.

---

### 5. **Addressing Node.js Ecosystem Concerns**
   - **Question:** Will using Zsh or macOS-specific setups lead to bugs in Node.js or associated tools?  
     - **Answer:** No, the ecosystem (Node.js, npm, TypeScript, etc.) should work seamlessly regardless of shell choice. Issues are more likely related to configuration rather than fundamental incompatibility.
   - **Workaround for Sass Needs:**  
     If your goal is simply to get Sass working, consider using `dart-sass`, which has fewer dependencies and works as a standalone tool:  
     ```bash
     brew install sass/sass/sass
     sass --version
     ```

---

### Final Thoughts  
It’s clear that the documentation and installation guides need a refresh to reflect modern development environments and tools. You are right to advocate for explicitness and clarity—it saves time and frustration for everyone. I recommend raising these concerns as an issue on the Node.js or NVM GitHub repositories for visibility and potential action.

**Comment by ljharb** [comment]
@MalikRumi feel free to file an issue on https://github.com/nvm-sh/nvm about this, but the documentation already says `bash`, and you should never feel comfortable blindly substituting one shell for another without first confirming compatibility. while nvm itself works fine on zsh, the install script only works on bash, and you can run the install command in zsh as long as you don't change it.

**Reply by hernandemonteiro** [reply]
i think that...
It’s not about the shell; the issue is that the install script calls `python` instead of `python3`. This can cause problems on systems where `python` isn’t mapped to Python 3.

**Reply by ljharb** [reply]
ah, that's part of node's own installation, which has nothing to do with nvm.
